exporters/otlp/src/otlp_file_client.cc

```diff
@@ -0,0 +1,1560 @@
+      }
+      if (background_flush_thread && background_flush_thread->joinable())
+      {
+        background_flush_thread->detach();
+      }
+    }));
+  }
+
+private:
+  OtlpFileClientFileSystemOptions options_;
+
+  struct FileStats
+  {
+    std::atomic<bool> is_shutdown;
+    std::size_t rotate_index;
+    std::size_t written_size;
+    std::size_t left_flush_record_count;
+    std::shared_ptr<std::ofstream> current_file;
+    std::mutex file_lock;
+    std::time_t last_checkpoint;
+    std::string file_path;
+    std::atomic<std::size_t> record_count;
+    std::atomic<std::size_t> flushed_record_count;
+
+    std::unique_ptr<std::thread> background_flush_thread;
+    std::mutex background_thread_lock;
+    std::mutex background_thread_waker_lock;
+    std::condition_variable background_thread_waker_cv;
+    std::mutex background_thread_waiter_lock;
+    std::condition_variable background_thread_waiter_cv;
+  };
+  std::shared_ptr<FileStats> file_;
+
+  std::atomic<bool> is_initialized_;
+  std::time_t check_file_path_interval_;
+};
+
+OtlpFileClient::OtlpFileClient(OtlpFileClientOptions &&options)
+    : is_shutdown_(false), options_(options)
+{
+  if (nostd::holds_alternative<OtlpFileClientFileSystemOptions>(options_.backend_options))
+  {
+    backend_ = std::make_shared<OtlpFileBackend>(
+        nostd::get<OtlpFileClientFileSystemOptions>(options_.backend_options));
+  }
+}
+
+OtlpFileClient::~OtlpFileClient()
+{
+  if (!IsShutdown())
+  {
+    Shutdown();
+  }
+}
+
+// ----------------------------- File Client methods ------------------------------
+opentelemetry::sdk::common::ExportResult OtlpFileClient::Export(
+    const google::protobuf::Message &message,
+    std::size_t record_count) noexcept
+{
+  if (is_shutdown_)
+  {
+    return ::opentelemetry::sdk::common::ExportResult::kFailure;
+  }
+
+  nlohmann::json json_request;
+  // Convert from proto into json object
+  ConvertGenericMessageToJson(json_request, message);
+
+  std::string post_body_json =
+      json_request.dump(-1, ' ', false, nlohmann::detail::error_handler_t::replace);
+  if (options_.console_debug)
+  {
+    OTEL_INTERNAL_LOG_DEBUG("[OTLP FILE Client] Write body(Json)" << post_body_json);
+  }
+
+  if (backend_)
+  {
+    backend_->Export(post_body_json, record_count);
+    return ::opentelemetry::sdk::common::ExportResult::kSuccess;
+  }
+  else if (nostd::holds_alternative<std::reference_wrapper<std::ostream>>(options_.backend_options))
+  {
+    std::ostream &out = nostd::get<std::reference_wrapper<std::ostream>>(options_.backend_options);
+    out.write(post_body_json.data(), post_body_json.size());
+    out.write("\n", 1);
+  }
+  else
+  {
+    return ::opentelemetry::sdk::common::ExportResult::kFailure;
+  }
+
+  return opentelemetry::sdk::common::ExportResult::kSuccess;
+}
+
+bool OtlpFileClient::ForceFlush(std::chrono::microseconds timeout) noexcept
+{
+  if (backend_)
+  {
+    return backend_->ForceFlush(timeout);
+  }
+
+  return true;
+}
```

Comment:
I think we should add the force flush mechanism for the std::ostream case as well. In case the user calls `TracerProvider::ForceFlush`, we should add something like this:

```
std::ostream &out = nostd::get<std::reference_wrapper<std::ostream>>(options_.backend_options);
out.flush();
```

Reply:
Good point. I didn't implement it before because some ostreams is not thread-safety and we must ignore `timeout`.
It may be more important to keep it right. I will add flush for ostream later.